staking/slash/double-sign.go

```diff
@@ -1,52 +1,341 @@
-		Signature *bls.Sign
-	} `json:"double-signed"`
-	Beneficiary common.Address // the reporter who will get rewarded
-}
-
-// NewRecord ..
-func NewRecord(
-	offender shard.BlsPublicKey,
-	signedHeader, doubleSignedHeader *block.Header,
-	signedSignature, doubleSignedSignature *bls.Sign,
-	beneficiary common.Address,
-) Record {
-	r := Record{}
-	r.Offender = offender
-	r.Signed.Header = signedHeader
-	r.Signed.Signature = signedSignature
-	r.DoubleSigned.Header = doubleSignedHeader
-	r.DoubleSigned.Signature = doubleSignedSignature
-	r.Beneficiary = beneficiary
-	return r
-}
-
-// TODO(Edgar) Implement Verify and Apply
+	// the reporter who will get rewarded
+	Evidence Evidence       `json:"evidence"`
+	Reporter common.Address `json:"reporter"`
+	Offender common.Address `json:"offender"`
+}
+
+// Application ..
+type Application struct {
+	TotalSlashed, TotalSnitchReward *big.Int
+}
+
+func (a *Application) String() string {
+	s, _ := json.Marshal(a)
+	return string(s)
+}
+
+// MarshalJSON ..
+func (e Evidence) MarshalJSON() ([]byte, error) {
+	return json.Marshal(struct {
+		Moment
+		ConflictingBallots
+		ProposalHeader string `json:"header"`
+	}{e.Moment, e.ConflictingBallots, e.ProposalHeader.String()})
+}
+
+// Records ..
+type Records []Record
+
+func (r Records) String() string {
+	s, _ := json.Marshal(r)
+	return string(s)
+}
+
+// MarshalJSON ..
+func (r Record) MarshalJSON() ([]byte, error) {
+	reporter, offender :=
+		common2.MustAddressToBech32(r.Reporter),
+		common2.MustAddressToBech32(r.Offender)
+	return json.Marshal(struct {
+		Evidence         Evidence `json:"evidence"`
+		Beneficiary      string   `json:"beneficiary"`
+		AddressForBLSKey string   `json:"offender"`
+	}{r.Evidence, reporter, offender})
+}
+
+func (e Evidence) String() string {
+	s, _ := json.Marshal(e)
+	return string(s)
+}
+
+func (r Record) String() string {
+	s, _ := json.Marshal(r)
+	return string(s)
+}
+
+// CommitteeReader ..
+type CommitteeReader interface {
+	ReadShardState(epoch *big.Int) (*shard.State, error)
+}
 
 // Verify checks that the signature is valid
-func Verify(candidate *Record) error {
+func Verify(chain CommitteeReader, candidate *Record) error {
+	first, second :=
+		candidate.Evidence.AlreadyCastBallot,
+		candidate.Evidence.DoubleSignedBallot
+	if shard.CompareBlsPublicKey(first.SignerPubKey, second.SignerPubKey) != 0 {
+		k1, k2 := first.SignerPubKey.Hex(), second.SignerPubKey.Hex()
+		return errors.Wrapf(
+			errBLSKeysNotEqual, "%s %s", k1, k2,
+		)
+	}
+	superCommittee, err := chain.ReadShardState(candidate.Evidence.Epoch)
+
+	if err != nil {
+		return err
+	}
+
+	subCommittee := superCommittee.FindCommitteeByID(
+		candidate.Evidence.ShardID,
+	)
+
+	if subCommittee == nil {
+		return errors.Wrapf(
+			errShardIDNotKnown, "given shardID %d", candidate.Evidence.ShardID,
+		)
+	}
+
+	if _, err := subCommittee.AddressForBLSKey(second.SignerPubKey); err != nil {
+		return err
+	}
+	// TODO need to finish this implementation
```

Comment:
Also need to check the validator is not slashed/banned already, to prevent repeated acceptance of same slash.

Reply:
This is not done yet.